Authorization -> Evaluate: always returns "No search results"

### Before reporting an issue

- [x] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

admin/ui

### Describe the bug

I'm trying to set up an application that's only accessible if you're part of a certain realm group, and to test it with the `Evaluate` function in `Clients -> Authorization -> Evaluate`. But this issue happens even on clients where I have an allow-all permission.

In the applications the policies **are enforced**. It's just the admin ui always gives me the same result when running evaluate. I'm not sure what to put in `roles`. It seems to have been made mandatory in some update to Keycloak.

## Allow All

1. Create Policy

![Image](https://github.com/user-attachments/assets/5db92cb6-f08d-4ed6-96bc-3c84e38358a1)

![Image](https://github.com/user-attachments/assets/2dbfd359-32fe-4e2a-8c10-b33c15b263bd)

2. Create Permission

![Image](https://github.com/user-attachments/assets/1a7ab1a9-4fdc-4bb9-b26a-fa9317b7c9ca)

![Image](https://github.com/user-attachments/assets/882f19a2-cfe8-4c35-a2be-a7a6932e10c9)

3. Evaluate

![Image](https://github.com/user-attachments/assets/4b1af849-f729-46bf-a7fe-78bc83a4dbd2)

4. Results are "No search results". This page used to tell me whether this user has access or not

![Image](https://github.com/user-attachments/assets/45ea26ac-59d3-4e24-ac47-cfdc8a7c3922)

## Based on group

![Image](https://github.com/user-attachments/assets/8ec38287-674e-4e2f-ac5b-4dd72e737b91)

![Image](https://github.com/user-attachments/assets/4f408e03-c1ff-4377-950a-248ce5f3bd14)

![Image](https://github.com/user-attachments/assets/804c894b-1614-4983-bf76-123c01e1fa50)

![Image](https://github.com/user-attachments/assets/8bdf2168-a2bf-4fbf-9212-9c1f34c7e1c7)

![Image](https://github.com/user-attachments/assets/cfe9a45f-7df3-4ea1-8346-ef5619def553)

![Image](https://github.com/user-attachments/assets/b19176f1-f89b-4974-a38a-b4536046ba4f)

### Version

26.2.5

### Regression

- [ ] The issue is a regression

### Expected behavior

I expect it to tell me whether this user has access to the client/application. In older versions of Keycloak this was the case. Also "roles" was not mandatory.

### Actual behavior

Always returns "No search results" 

### How to Reproduce?

See screenshots

### Anything else?

_No response_

Running 26.4.0

Interestingly if I turn off the new admin permissions I get an error with my `freshrss` client:  `Could not evaluate due to: server_error. Error while evaluating permissions.`

Other clients seem to return the evaluation result with admin permissions turned off.

Error I get with the `freshrss` client:

```
keycloak           | 2025-10-17 22:04:28,209 ERROR [org.keycloak.authorization.admin.PolicyEvaluationService] (executor-thread-6) Error while evaluating permissions: java.lang.RuntimeException: Failed to evaluate permissions
keycloak           |    at org.keycloak.authorization.policy.evaluation.DecisionPermissionCollector.onError(DecisionPermissionCollector.java:189)
keycloak           |    at org.keycloak.authorization.permission.evaluator.UnboundedPermissionEvaluator.evaluate(UnboundedPermissionEvaluator.java:53)
keycloak           |    at org.keycloak.authorization.admin.PolicyEvaluationService.evaluate(PolicyEvaluationService.java:156)
keycloak           |    at org.keycloak.authorization.admin.PolicyEvaluationService.evaluate(PolicyEvaluationService.java:139)
keycloak           |    at org.keycloak.authorization.admin.PolicyEvaluationService$quarkusrestinvoker$evaluate_456fb9dc423e56412cc98b52fc9ce42b2a4ff5b0.invoke(Unknown Source)
keycloak           |    at org.jboss.resteasy.reactive.server.handlers.InvocationHandler.handle(InvocationHandler.java:29)
keycloak           |    at io.quarkus.resteasy.reactive.server.runtime.QuarkusResteasyReactiveRequestContext.invokeHandler(QuarkusResteasyReactiveRequestContext.java:183)
keycloak           |    at org.jboss.resteasy.reactive.common.core.AbstractResteasyReactiveContext.run(AbstractResteasyReactiveContext.java:147)
keycloak           |    at io.quarkus.vertx.core.runtime.VertxCoreRecorder$15.runWith(VertxCoreRecorder.java:645)
keycloak           |    at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
keycloak           |    at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
keycloak           |    at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
keycloak           |    at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
keycloak           |    at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:11)
keycloak           |    at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:11)
keycloak           |    at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
keycloak           |    at java.base/java.lang.Thread.run(Thread.java:1583)
keycloak           | Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.CharSequence.length()" because "this.text" is null
keycloak           |    at java.base/java.util.regex.Matcher.getTextLength(Matcher.java:1808)
keycloak           |    at java.base/java.util.regex.Matcher.reset(Matcher.java:461)
keycloak           |    at java.base/java.util.regex.Matcher.<init>(Matcher.java:256)
keycloak           |    at java.base/java.util.regex.Pattern.matcher(Pattern.java:1180)
keycloak           |    at org.keycloak.utils.JsonUtils.hasPath(JsonUtils.java:66)
keycloak           |    at org.keycloak.authorization.policy.provider.regex.RegexPolicyProvider.getClaimValue(RegexPolicyProvider.java:82)
keycloak           |    at org.keycloak.authorization.policy.provider.regex.RegexPolicyProvider.evaluate(RegexPolicyProvider.java:61)
keycloak           |    at org.keycloak.authorization.policy.provider.permission.AbstractPermissionProvider.evaluate(AbstractPermissionProvider.java:62)
```

--- a/services/src/main/java/org/keycloak/authorization/admin/PolicyEvaluationService.java
+++ b/services/src/main/java/org/keycloak/authorization/admin/PolicyEvaluationService.java
@@ -150,7 +150,7 @@ public class PolicyEvaluationService {
         EvaluationDecisionCollector decision = new EvaluationDecisionCollector(authorization, resourceServer, request);
 
         if (permissions.isEmpty()) {
-            if (AdminPermissionsSchema.SCHEMA.isAdminPermissionsEnabled(authorization.getRealm())) {
+            if (AdminPermissionsSchema.SCHEMA.isAdminPermissionClient(authorization.getRealm(), resourceServer.getId())) {
                 return decision;
             }
             return authorization.evaluators().from(evaluationContext, resourceServer, request).evaluate(decision);

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/authz/PolicyEvaluationTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/authz/PolicyEvaluationTest.java
@@ -16,8 +16,17 @@
  */
 package org.keycloak.testsuite.authz;
 
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+
+import jakarta.ws.rs.core.Response;
 import org.junit.Assert;
 import org.junit.Test;
+import org.keycloak.admin.client.resource.AuthorizationResource;
+import org.keycloak.admin.client.resource.ClientResource;
+import org.keycloak.admin.client.resource.ClientsResource;
+import org.keycloak.admin.client.resource.RealmResource;
 import org.keycloak.authorization.AuthorizationProvider;
 import org.keycloak.authorization.Decision.Effect;
 import org.keycloak.authorization.attribute.Attributes;
@@ -40,15 +49,21 @@ import org.keycloak.models.utils.KeycloakModelUtils;
 import org.keycloak.protocol.oidc.OIDCLoginProtocol;
 import org.keycloak.protocol.oidc.mappers.GroupMembershipMapper;
 import org.keycloak.protocol.oidc.mappers.OIDCAttributeMapperHelper;
+import org.keycloak.representations.idm.ClientRepresentation;
 import org.keycloak.representations.idm.GroupRepresentation;
 import org.keycloak.representations.idm.ProtocolMapperRepresentation;
 import org.keycloak.representations.idm.RealmRepresentation;
+import org.keycloak.representations.idm.UserRepresentation;
 import org.keycloak.representations.idm.authorization.JSPolicyRepresentation;
 import org.keycloak.representations.idm.authorization.Logic;
 import org.keycloak.representations.idm.authorization.Permission;
+import org.keycloak.representations.idm.authorization.PolicyEvaluationRequest;
+import org.keycloak.representations.idm.authorization.PolicyEvaluationResponse;
 import org.keycloak.representations.idm.authorization.ResourcePermissionRepresentation;
+import org.keycloak.representations.idm.authorization.ResourceRepresentation;
 import org.keycloak.representations.idm.authorization.ScopePermissionRepresentation;
 import org.keycloak.representations.idm.authorization.TimePolicyRepresentation;
+import org.keycloak.representations.idm.authorization.UserPolicyRepresentation;
 import org.keycloak.testsuite.util.ClientBuilder;
 import org.keycloak.testsuite.util.GroupBuilder;
 import org.keycloak.testsuite.util.RealmBuilder;
@@ -145,7 +160,7 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
         PolicyProvider provider = authorization.getProvider(policy.getType());
         DefaultEvaluation evaluation = createEvaluation(session, authorization, resourceServer, policy);
         provider.evaluate(evaluation);
-        Assert.assertEquals(Effect.PERMIT, evaluation.getEffect());
+        assertEquals(Effect.PERMIT, evaluation.getEffect());
 
         // lets now override the context to use a time that exceeds the time that was set in the policy.
         long contextTime = System.currentTimeMillis() + 5400000;
@@ -153,7 +168,7 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
         attributes.put("kc.time.date_time", Arrays.asList(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(new Date(contextTime))));
         evaluation = createEvaluation(session, authorization, null, resourceServer, policy, attributes);
         provider.evaluate(evaluation);
-        Assert.assertEquals(Effect.DENY, evaluation.getEffect());
+        assertEquals(Effect.DENY, evaluation.getEffect());
     }
 
     @Test
@@ -192,7 +207,7 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
 
         provider.evaluate(evaluation);
 
-        Assert.assertEquals(Effect.PERMIT, evaluation.getEffect());
+        assertEquals(Effect.PERMIT, evaluation.getEffect());
 
         policyRepresentation = new JSPolicyRepresentation();
         policyRepresentation.setName("allow-user-in-group-path-a-policy");
@@ -204,7 +219,7 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
 
         provider.evaluate(evaluation);
 
-        Assert.assertEquals(Effect.PERMIT, evaluation.getEffect());
+        assertEquals(Effect.PERMIT, evaluation.getEffect());
 
         policyRepresentation = new JSPolicyRepresentation();
         policyRepresentation.setName("allow-user-in-group-path-b-policy");
@@ -228,7 +243,7 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
 
         provider.evaluate(evaluation);
 
-        Assert.assertEquals(Effect.PERMIT, evaluation.getEffect());
+        assertEquals(Effect.PERMIT, evaluation.getEffect());
 
         policyRepresentation = new JSPolicyRepresentation();
         policyRepresentation.setName("allow-alice-in-group-path-a-policy");
@@ -240,7 +255,7 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
 
         provider.evaluate(evaluation);
 
-        Assert.assertEquals(Effect.PERMIT, evaluation.getEffect());
+        assertEquals(Effect.PERMIT, evaluation.getEffect());
 
         policyRepresentation = new JSPolicyRepresentation();
         policyRepresentation.setName("allow-alice-in-group-path-a-no-parent-policy.js");
@@ -302,7 +317,7 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
 
         provider.evaluate(evaluation);
 
-        Assert.assertEquals(Effect.PERMIT, evaluation.getEffect());
+        assertEquals(Effect.PERMIT, evaluation.getEffect());
 
         policyRepresentation = new JSPolicyRepresentation();
         policyRepresentation.setId(null);
@@ -341,7 +356,7 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
 
         provider.evaluate(evaluation);
 
-        Assert.assertEquals(Effect.PERMIT, evaluation.getEffect());
+        assertEquals(Effect.PERMIT, evaluation.getEffect());
 
         policyRepresentation = new JSPolicyRepresentation();
         policyRepresentation.setName("allow-trinity-in-client-role-b-policy");
@@ -379,7 +394,7 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
 
         provider.evaluate(evaluation);
 
-        Assert.assertEquals(Effect.PERMIT, evaluation.getEffect());
+        assertEquals(Effect.PERMIT, evaluation.getEffect());
 
         policyRepresentation = new JSPolicyRepresentation();
         policyRepresentation.setType("script-scripts/allow-child-group-in-role-policy.js");
@@ -418,7 +433,7 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
 
         provider.evaluate(evaluation);
 
-        Assert.assertEquals(Effect.PERMIT, evaluation.getEffect());
+        assertEquals(Effect.PERMIT, evaluation.getEffect());
     }
 
     @Test
@@ -444,7 +459,7 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
 
         provider.evaluate(evaluation);
 
-        Assert.assertEquals(Effect.PERMIT, evaluation.getEffect());
+        assertEquals(Effect.PERMIT, evaluation.getEffect());
     }
 
     @Test
@@ -470,7 +485,7 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
 
         provider.evaluate(evaluation);
 
-        Assert.assertEquals(Effect.PERMIT, evaluation.getEffect());
+        assertEquals(Effect.PERMIT, evaluation.getEffect());
     }
 
     @Test
@@ -502,7 +517,7 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
 
         provider.evaluate(evaluation);
 
-        Assert.assertEquals(Effect.PERMIT, evaluation.getEffect());
+        assertEquals(Effect.PERMIT, evaluation.getEffect());
     }
 
     @Test
@@ -532,7 +547,7 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
 
         provider.evaluate(evaluation);
 
-        Assert.assertEquals(Effect.PERMIT, evaluation.getEffect());
+        assertEquals(Effect.PERMIT, evaluation.getEffect());
     }
 
     @Test
@@ -553,27 +568,35 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
 
         Policy policy = storeFactory.getPolicyStore().create(resourceServer, policyRepresentation);
 
-        Resource resource = storeFactory.getResourceStore().create(resourceServer, "Resource A", resourceServer.getClientId());
-        Scope scope = storeFactory.getScopeStore().create(resourceServer, "Scope A");
-
-        resource.updateScopes(new HashSet<>(Arrays.asList(scope)));
-
-        ResourcePermissionRepresentation permission = new ResourcePermissionRepresentation();
-
-        permission.setName("testCheckReadOnlyInstances permission");
-        permission.addPolicy(policy.getId());
-        permission.addResource(resource.getId());
-
-        storeFactory.getPolicyStore().create(resourceServer, permission);
-
-        session.getTransactionManager().commit();
-
-        PermissionEvaluator evaluator = authorization.evaluators().from(Arrays.asList(new ResourcePermission(resource, Arrays.asList(scope), resourceServer)), createEvaluationContext(session, Collections.emptyMap()));
-
         try {
-            evaluator.evaluate(resourceServer, null);
-            Assert.fail("Instances should be marked as read-only");
-        } catch (Exception ignore) {
+            Resource resource = storeFactory.getResourceStore().create(resourceServer, "Resource A", resourceServer.getClientId());
+            Scope scope = storeFactory.getScopeStore().create(resourceServer, "Scope A");
+
+            resource.updateScopes(new HashSet<>(Arrays.asList(scope)));
+
+            ResourcePermissionRepresentation permission = new ResourcePermissionRepresentation();
+
+            permission.setName("testCheckReadOnlyInstances permission");
+            permission.addPolicy(policy.getId());
+            permission.addResource(resource.getId());
+
+            storeFactory.getPolicyStore().create(resourceServer, permission);
+
+            session.getTransactionManager().commit();
+
+            PermissionEvaluator evaluator = authorization.evaluators().from(Arrays.asList(new ResourcePermission(resource, Arrays.asList(scope), resourceServer)), createEvaluationContext(session, Collections.emptyMap()));
+
+            try {
+                evaluator.evaluate(resourceServer, null);
+                Assert.fail("Instances should be marked as read-only");
+            } catch (Exception ignore) {
+            }
+        } finally {
+            KeycloakModelUtils.runJobInTransaction(session.getKeycloakSessionFactory(), session1 -> {
+                AuthorizationProvider authorization1 = session1.getProvider(AuthorizationProvider.class);
+                StoreFactory storeFactory1 = authorization1.getStoreFactory();
+                storeFactory1.getPolicyStore().delete(policy.getId());
+            });
         }
     }
 
@@ -621,7 +644,7 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
         PermissionEvaluator evaluator = authorization.evaluators().from(Arrays.asList(new ResourcePermission(resource, Arrays.asList(readScope, writeScope), resourceServer)), createEvaluationContext(session, Collections.emptyMap()));
         Collection<Permission> permissions = evaluator.evaluate(resourceServer, null);
 
-        Assert.assertEquals(0, permissions.size());
+        assertEquals(0, permissions.size());
     }
 
     private static DefaultEvaluation createEvaluation(KeycloakSession session, AuthorizationProvider authorization, ResourceServer resourceServer, Policy policy) {
@@ -663,5 +686,42 @@ public class PolicyEvaluationTest extends AbstractAuthzTest {
                 return baseAttributes;
             }
         };
+    }
+
+    @Test
+    public void testEvaluation() {
+        RealmResource realmApi = realmsResouce().realm("authz-test");
+        RealmRepresentation realm = realmApi.toRepresentation();
+        realm.setAdminPermissionsEnabled(true);
+        realmApi.update(realm);
+        getCleanup(realm.getRealm()).addCleanup(() -> realm.setAdminPermissionsEnabled(false));
+        ClientsResource clientsApi = realmApi.clients();
+        ClientRepresentation client = clientsApi.findByClientId("resource-server-test").get(0);
+        ClientResource clientApi = clientsApi.get(client.getId());
+        AuthorizationResource authorizationApi = clientApi.authorization();
+        UserRepresentation user = realmApi.users().search("marta").get(0);
+        ResourceRepresentation resource = new ResourceRepresentation(KeycloakModelUtils.generateId());
+        try (Response response = authorizationApi.resources().create(resource)) {
+            assertEquals(201,  response.getStatus());
+        }
+
+        UserPolicyRepresentation policy = new UserPolicyRepresentation();
+        policy.setName(KeycloakModelUtils.generateId());
+        policy.addUser(user.getId());
+        authorizationApi.policies().user().create(policy).close();
+
+        ResourcePermissionRepresentation permission = new ResourcePermissionRepresentation();
+        permission.setName(KeycloakModelUtils.generateId());
+        permission.addResource(resource.getName());
+        permission.addPolicy(policy.getName());
+        authorizationApi.permissions().resource().create(permission).close();
+
+        PolicyEvaluationRequest request = new PolicyEvaluationRequest();
+        request.setUserId(user.getId());
+        request.setClientId(client.getId());
+//        request.addResource(resource.getName());
+        PolicyEvaluationResponse result = authorizationApi.policies().evaluate(request);
+        assertNotNull(result.getResults());
+        assertFalse(result.getResults().isEmpty());
     }
 }

